### PR TITLE
Move Staging to Typer

### PR DESCRIPTION
**Pull request by nicolasstucki**
The main motivation for moving staging to typer is to support whitebox macros.

Using the ` <: T` return type notation.
```scala
/** Returns a Char if the string is of length 1 or a String otherwise */
inline def charOrString(inline str: String) <: Char | String = ${ impl(str) }
def impl(str: String) = if (str.length == 1) str.charAt(0).toExpr else str.toExpr
// Different file
val x: Char = charOrString("1")
val y: String = charOrString("123")
```

Modification:
* `Staging` is used in `Typer` to check contents of inline macros
* Expand macros in `Inliner` durring typing
* Keep track of quotation level in the context
* Lose support for inspection of any tree in the current compilation unit. With the exception of parameter bindings which can be inspected.

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Needs rebase on #5940

**Comment by OlivierBlanvillain** [comment]
@nicolasstucki Can you explain the motivation for this change?

**Comment by nicolasstucki** [comment]
Mainly to support whitebox macros

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+          mapOver(tp)
+      }
+    }
+  }
+
+  /** Check reference to `sym` for phase consistency, where `tp` is the underlying type
+   *  by which we refer to `sym`. If it is an inconsistent type try construct a healed type for it.
+   *
+   *  @return `None` if the phase is correct or cannot be healed
+   *          `Some(tree)` with the `tree` of the healed type tree for `~implicitly[quoted.Type[T]]`
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: ~ --> $

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+   */
+  private def checkSymLevel(sym: Symbol, tp: Type, pos: SourcePosition)(implicit ctx: Context): Option[Tree] = {
+    val isThis = tp.isInstanceOf[ThisType]
+    if (!isThis && !sym.is(Param) && sym.maybeOwner.isType)
+      None
+    else if (sym.exists && !sym.isStaticOwner && !levelOK(sym))
+      tryHeal(sym, tp, pos)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: The logic here is very hard to follow, to the point where I cannot be sure it is correct. 
A clearer formulation would be:
```
val needsCheck =
  tp.isInstanceOf[ThisType] || sym.is(Param) || sym.owner.isTerm
if (sym.exists && needsCheck && !sym.isStaticOwner && !levelOK(sym)) 
  tryHeal(sym, tp, pos)
else
  None
```
and it would be good to have further explanations in comments.

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 3]
```diff
@@ -185,6 +185,27 @@ object Applications {
       case _ => None
     }
   }
+
+  /** 1. If we are in an inline method but not in a nested quote, mark the inline method
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala]: The code seems to mark a macro also for quotes. So it seems comment and code disagree. Which is right?

File: compiler/src/dotty/tools/dotc/transform/Splicer.scala [hunk 4]
```diff
@@ -312,7 +312,13 @@ object Splicer {
 
     protected final def interpretTree(tree: Tree)(implicit env: Env): Result = tree match {
       case Apply(TypeApply(fn, _), quoted :: Nil) if fn.symbol == defn.QuotedExpr_apply =>
-        interpretQuote(quoted)
+        val quoted1 = quoted match {
+          case quoted: Ident if quoted.symbol.is(InlineProxy) && quoted.symbol.is(MethodFlag) => // inline proxy for by-name parameter
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/Splicer.scala]: Move comment to next line to shorten

Comment (odersky) [reply compiler/src/dotty/tools/dotc/transform/Splicer.scala]: You can just use `quoted.symbol.is(InlineMethod)`. Then you don't need to rename `Method` to `MethodFlag`.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 5]
```diff
@@ -1551,13 +1555,18 @@ class Typer extends Namer
...
+      PrepareInlineable.registerInlineInfo(sym, _ => rhs1)
+    }
 
     if (sym.isConstructor && !sym.isPrimaryConstructor)
       for (param <- tparams1 ::: vparamss1.flatten)
         checkRefsLegal(param, sym.owner, (name, sym) => sym.is(TypeParam), "secondary constructor")
 
-    assignType(cpy.DefDef(ddef)(name, tparams1, vparamss1, tpt1, rhs1), sym)
+    val ddef1 = assignType(cpy.DefDef(ddef)(name, tparams1, vparamss1, tpt1, rhs1), sym)
+    sym.defTree = ddef1
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: I wonder whether we should factor out the `defTree` setting, into a method in `tpd.TreeOps` like:
```
def setDefTree: tree.type = {
  val sym = tree.symbol
  if (sym.exists) sym.defTree = tree
  tree
}
```
We seem to have 4 uses in this change set, and maybe more elsewhere.

File: compiler/src/dotty/tools/dotc/typer/PrepareInlineable.scala [hunk 6]
```diff
@@ -249,4 +251,52 @@ object PrepareInlineable {
         em"inline unapply method can be rewritten only if its right hand side is a tuple (e1, ..., eN)",
         body.sourcePos)
   }
+
+  def checkInlineMacro(sym: Symbol, rhs: Tree, pos: SourcePosition)(implicit ctx: Context) = {
+    if (ctx.phase.isTyper) {
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/PrepareInlineable.scala]: Why not use `if (!ctx.isAfterTyper)` ?

File: compiler/src/dotty/tools/dotc/typer/PrepareInlineable.scala [hunk 7]
```diff
@@ -249,4 +251,52 @@ object PrepareInlineable {
...
+        sym.setFlag(Macro)
+        if (level == 0)
+          rhs match {
+            case InlineSplice(_) =>
+              new PCPCheckAndHeal(freshStagingContext).transform(rhs) // Ignore output, only check PCP
+            case _ =>
+              ctx.error(
+                """Malformed macro.
+                  |
+                  |Expected the ~ to be at the top of the RHS:
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/PrepareInlineable.scala]: ~ --> $ (better grep for it!)

File: compiler/src/dotty/tools/dotc/transform/Splicer.scala (hunk 4 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/Splicer.scala]: Actually this is is not an `InlineMethod` which is `Inline | Method`. This one is `InlineProxy | Method` which could mark the `def` binding for by-name parameters. I will create `InlineByNameProxy` to make the intent clear.

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+import dotty.tools.dotc.typer.Inliner
+
+import scala.collection.mutable
+import dotty.tools.dotc.util.SourcePosition
+
+import scala.annotation.constructorOnly
+
+/** Checks that the Phase Consistency Principle (PCP) holds and heals types.
+ *
+ *  Type healing consists in transforming a phase inconsistent type `T` into `implicitly[Type[T]].unary_~`.
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: Changed

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    else {
+      assert(!enclosingInlineds.nonEmpty, "unexpanded macro")
+      assert(ctx.owner.isInlineMethod)
+      if (Splicer.canBeSpliced(splice.qualifier)) { // level 0 inside an inline definition
+        transform(splice.qualifier)(spliceContext) // Just check PCP
+        splice
+      }
+      else { // level 0 inside an inline definition
+        ctx.error(
+          "Malformed macro call. The contents of the $ must call a static method and arguments must be quoted or inline.",
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: Changed

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    val tp = checkType(tree.sourcePos).apply(tree.tpe.widenTermRefExpr)
+    tree.cast(tp).withSpan(tree.span)
+  }
+
+  /** If `tree` refers to a locally defined symbol (either directly, or in a pickled type),
+   *  check that its staging level matches the current level. References to types
+   *  that are phase-incorrect can still be healed as follows:
+   *
+   *  If `T` is a reference to a type at the wrong level, try to heal it by replacing it with
+   *  `~implicitly[quoted.Type[T]]`.
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: Changed

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: Changed

File: compiler/src/dotty/tools/dotc/typer/PrepareInlineable.scala (hunk 7 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/PrepareInlineable.scala]: Changed

File: compiler/src/dotty/tools/dotc/typer/PrepareInlineable.scala (hunk 6 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/PrepareInlineable.scala]: Yes, that was the intention

File: compiler/src/dotty/tools/dotc/typer/Applications.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Applications.scala]: The comment is right. We used to overapproximate and mark them all as macros which was not ideal. I added a commit fixing this.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 5 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: It seems worth it. I found 6 instances where it could be used. I will add a commit with this change.

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: I partially refactored it.

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+    }
+  }
+
+
+  /** Add cast to force boundaries where T and ~t (an alias of T) are used to ensure PCP.
+   *  '{   ~(...: T)  }  -->  '{   ~(...: T).asInstanceOf[T]  } --> '{   ~(...: T).asInstanceOf[~t]  }
+   */
+  protected def addSpliceCast(tree: Tree)(implicit ctx: Context): Tree = {
+    val tp = checkType(tree.sourcePos).apply(tree.tpe.widenTermRefExpr)
+    tree.cast(tp).withSpan(tree.span)
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: It does not always need to cast, but `ensureConforms` would not work here because we always cast to a type that is equal modulo spliced type aliases. 

A recording of the splice into something that is not the type alias `$splice` in `quoted.Type` could avoid this cast altogether. I intend to do this in a future PR, I will need it to hide internal encoding from the user and in #5297.

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala (hunk 2 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: `shouldtryHealable` can be a `def`, since it is only called after 3 other conditions. 
It would be good to add a short comment to each of the 3 lines of shouldTryHealable, indicating why we
should try to heal in this case.

**Comment by odersky** [comment]
I only have one more comment on the last commit. Otherwise it's good.

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: After playing around with this condition I figured out that is was checking is the symbol is a reference to a class (but not a `this.type`). I updated the code accordingly.

**Comment by nicolasstucki** [comment]
Rebased

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala (hunk 2 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: Ah, that clarifies things!
